services/galley/src/Galley/API/Util.hs

```diff
@@ -342,3 +346,57 @@ runFederated :: forall (c :: Component) a. Domain -> FederatorClient c (ExceptT
 runFederated remoteDomain rpc = do
   runExceptT (executeFederated remoteDomain rpc)
     >>= either (throwM . federationErrorToWai) pure
+
+-- | Notify local users and bots of being added to a conversation
+notifyLocals :: [InternalMember UserId] -> [BotMember] -> Event -> UserId -> ConnId -> [InternalMember UserId] -> Galley ()
+notifyLocals existingLocals bots e usr conn newLocals = do
+  let allMembers = nubOrdOn memId (newLocals <> existingLocals)
+  for_ (newPush ListComplete usr (ConvEvent e) (recipient <$> allMembers)) $ \p ->
+    push1 $ p & pushConn ?~ conn
+  void . forkIO $ void $ External.deliver (bots `zip` repeat e)
+
+-- | Notify remote users of being added to a conversation
+notifyRemotes :: [RemoteMember] -> UserId -> UTCTime -> Data.Conversation -> [LocalMember] -> [RemoteMember] -> Galley ()
+notifyRemotes existingRemotes usr now c lmm rmm = do
+  localDomain <- viewFederationDomain
+  let mm = catMembers localDomain lmm rmm
+      qcnv = Qualified (Data.convId c) localDomain
+      qusr = Qualified usr localDomain
+  -- FUTUREWORK: parallelise federated requests
+  traverse_ (uncurry (updateRemoteConversations now mm qusr qcnv))
+    . Map.assocs
+    . partitionQualified
+    . nubOrd
+    . map (unTagged . rmId)
+    $ rmm <> existingRemotes
+  where
+    catMembers ::
+      Domain ->
+      [LocalMember] ->
+      [RemoteMember] ->
+      [(Qualified UserId, RoleName)]
+    catMembers localDomain ls rs =
+      map (((`Qualified` localDomain) . memId) &&& memConvRoleName) ls
+        <> map ((unTagged . rmId) &&& rmConvRoleName) rs
+
+updateRemoteConversations ::
+  UTCTime ->
+  [(Qualified UserId, RoleName)] ->
+  Qualified UserId ->
+  Qualified ConvId ->
+  Domain ->
+  [UserId] ->
+  Galley ()
+updateRemoteConversations now uids orig cnv domain others = do
+  let cmu =
+        ConversationMemberUpdate
+          { cmuTime = now,
+            cmuOrigUserId = orig,
+            cmuConvId = cnv,
+            cmuAlreadyPresentUsers = others,
+            cmuUsersAdd = uids,
+            cmuUsersRemove = []
+          }
+  let rpc = FederatedGalley.updateConversationMemberships FederatedGalley.clientRoutes cmu
+  runExceptT (Federation.executeFederated domain rpc)
+    >>= either (throwM . federationErrorToWai) pure
```

Comment:
While you're at it, could you make use of the `runFederated` convenience function defined just above?

Reply:
I implemented this.